### PR TITLE
feat(ppo): in-model value head for Megatron PPO

### PR DESCRIPTION
**Pull request by bg51717**
# What does this PR do ?

Re-architects the Megatron PPO value model to use an in-model `LinearForLastLayer` output head instead of an external `ValueHead`, so the head inherits Megatron's gradient sync, optimizer, dist-checkpointing, PP, and SP — and adds value-model Sequence Parallelism, Dynamic Batching, and Pipeline Parallelism support.

Related issue: https://github.com/NVIDIA-NeMo/RL/issues/2687

# Usage
The value head is now the model's `output_layer`, so value-model parallelism/batching is configured exactly like the policy — e.g. in a PPO recipe:

```yaml
value:
  megatron_cfg:
    tensor_model_parallel_size: 2
    sequence_parallel: true          # now supported
    pipeline_model_parallel_size: 2  # now supported
  dynamic_batching:
    enabled: true                    # now supported
```

# Before your PR is "Ready for review"

**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Test
Tests run (all pass):

- **Unit** (`tests/unit/models/value/test_megatron_value_worker.py`, mcore, 2 GPU): init / `get_values` / train / checkpoint round-trip across DP, TP, SP, and dynamic batching, plus numerical-equivalence parity tests for SP and dynamic batching. 13 passed.
- **Functional** (`tests/functional/ppo_megatron.sh`, Qwen2.5-0.5B): passes. Confirms the base-checkpoint load fix — tied-embedding models no longer crash now that the value head is excluded from the base load.
- **Convergence** (Qwen2.5-1.5B, GSM8K, 1n8g, 40 steps), train reward / val accuracy:

baseline (value TP2): 0.84 / 0.74
 截屏2026-06-15 21 15 35

value SP + dynamic batching: 0.85 / 0.75
截屏2026-06-15 21 17 22

value PP2: 0.84 / 0.73
截屏2026-06-15 21 20 22  

  All converge identically — the new value-model features are numerically sound.
- **Checkpoint resume**: value head saved to / loaded from the dist-checkpoint (no separate sidecar), verified under both DP and PP2; resumed runs reach the same step-40 metrics as the baseline.
image

### CODE REVIEW

File: nemo_rl/models/value/workers/megatron_value_worker.py [hunk 1]
```diff
@@ -782,58 +651,29 @@ def forward_step_fn(
...
-                    hidden_states = captured_hidden["hidden_states"]
-                    # Megatron-Core always uses [S, B, H] layout; transpose to [B, S, H]
-                    hidden_states = hidden_states.transpose(0, 1).contiguous()
-                    values = self.value_head(hidden_states).squeeze(-1)  # [B, S]
+                # On the last PP stage the value head returns [B, S, 1] per-token
+                # values (sequence-parallel shards already gathered inside the
+                # head). Earlier PP stages return intermediate hidden states; their
+                # collection output is discarded (values are broadcast from the
+                # last stage below).
+                if is_pipeline_last_stage(ignore_virtual=True):
```
Comment (yuki-97) [inline nemo_rl/models/value/workers/megatron_value_worker.py:707]: The train path at line 130 uses `ignore_virtual=False` to gate the squeeze+shift on the head-owning vp chunk under VPP, but this `collection_fn` site uses `ignore_virtual=True`. Under VPP, non-head-owning vp chunks on the last PP rank would enter this branch on raw hidden states. Either match line 130's gate, or assert `virtual_pipeline_model_parallel_size in (None, 1)` in `__init__` since VPP isn't tested for the value worker.

File: tests/unit/models/value/test_megatron_value_worker.py [hunk 2]
```diff
@@ -297,6 +331,89 @@ def test_value_worker_train_step(value_setup):
     value.finish_training()
 
 
+@pytest.mark.hf_gated
+@pytest.mark.timeout(420)
+@pytest.mark.parametrize(
+    ("tp", "feature_updates"),
+    [
+        (2, {"sequence_parallel": True}),
+        (1, {"dynamic_batching": True}),
```
Comment (yuki-97) [inline tests/unit/models/value/test_megatron_value_worker.py:342]: This equivalence test covers SP and dynamic_batching but not PP, even though "Add MCore PPO PP" is the PR headline. `_apply_config_updates` also doesn't accept a `pipeline_model_parallel_size` key. PP correctness is only validated by the release-tier `valuetp2pp2` recipe today — a `(tp=1, pp=2, {})` case here would catch regressions at PR time.

File: nemo_rl/models/value/workers/megatron_value_worker.py [hunk 3]
```diff
@@ -352,25 +318,22 @@ def __init__(
...
-                if hasattr(model_module, "module"):
-                    model_module = model_module.module
-                if hasattr(model_module, "output_layer"):
-                    for param in model_module.output_layer.parameters():
-                        param.requires_grad_(False)
+        # The value head is an independent hidden->1 head, not tied to the input
+        # embedding. Untie before building so PP>1 does not set up an embedding/
+        # output-weight grad all-reduce that mismatches the [1, hidden] head and hangs.
+        self.megatron_cfg.model.share_embeddings_and_output_weights = False
+        # Replace output_layer with a hidden->1 value head before DDP wrapping, so grad
```
Comment (yuki-97) [inline nemo_rl/models/value/workers/megatron_value_worker.py:365]: Old `ValueHead` was `nn.Linear(hidden, 1, bias=True)`; `LinearForLastLayer` upstream hard-codes `bias=False`. The 40-step GSM8K release test likely doesn't surface the missing offset (already near zero-mean), but tasks with non-zero-mean baselines could regress. Intentional? If so please note in the PR description; otherwise upstream a `bias=True` option.

File: examples/configs/recipes/llm/ppo-qwen2.5-1.5b-gsm8k-1n8g-megatron-valuetp2sp-dynbatch.yaml
Comment (yuki-97) [inline examples/configs/recipes/llm/ppo-qwen2.5-1.5b-gsm8k-1n8g-megatron-valuetp2sp-dynbatch.yaml:1]: I think we can merge the new added two to one with `valuetp2sppp2-dynbatch` and replace this one `ppo-qwen2.5-1.5b-gsm8k-1n8g-megatron`. and still put it in nightly instead of release.

**Comment by yuki-97** [comment]
@cuichenx @yaoyu-33 , could you help to take a review as well?

**Review by yaoyu-33 — DISMISSED** [review]

File: nemo_rl/models/value/workers/megatron_value_worker.py (hunk 1 above)
Comment (bg51717) [reply nemo_rl/models/value/workers/megatron_value_worker.py:707]: Added assert `self.megatron_cfg.model.virtual_pipeline_model_parallel_size in (None, 1)` at worker init.
Both `forward_step_value` and `get_values` now use `ignore_virtual=True`.

File: tests/unit/models/value/test_megatron_value_worker.py (hunk 2 above)
Comment (bg51717) [reply tests/unit/models/value/test_megatron_value_worker.py:342]: Added. `_apply_config_updates` now accepts `pipeline_model_parallel_size`, and `test_value_worker_parallelism_equivalence` has a (tp=1, pp=2) case asserting get_values matches across PP degrees. Verified on the 2-GPU mcore runner.

File: nemo_rl/models/value/workers/megatron_value_worker.py (hunk 3 above)
Comment (bg51717) [reply nemo_rl/models/value/workers/megatron_value_worker.py:365]: `bias=False` is hardcoded in Megatron-Bridge's `LinearForLastLayer` :
https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/554c7b9324225aa863eee52e8b8fdde7abced2b1/src/megatron/bridge/training/utils/train_utils.py#L72
I think `bias=False` does not have much impact on the final result.

File: examples/configs/recipes/llm/ppo-qwen2.5-1.5b-gsm8k-1n8g-megatron-valuetp2sp-dynbatch.yaml
Comment (bg51717) [reply examples/configs/recipes/llm/ppo-qwen2.5-1.5b-gsm8k-1n8g-megatron-valuetp2sp-dynbatch.yaml:1]: PP and dynamic batching are mutually exclusive on the value model. Both `lm_value.py` and `lm_policy.py` `assert pipeline_model_parallel_size == 1` when `dynamic batching` is enabled ("Dynamic batching is only supported for single pipeline parallel stage").
So I kept the two recipes separate (each exercises one feature).

截屏2026-06-18 00 36 11

**Review by yuki-97 — APPROVED** [review]
